octoprint_octolight/__init__.py

```diff
@@ -39,69 +64,96 @@ def get_assets(self):
 		self._logger.info("OctoLight started, listening for GET request")
-		self._logger.info("Light pin: {}, inverted_input: {}".format(
-			self._settings.get(["light_pin"]),
-			self._settings.get(["inverted_output"])
+		self._logger.debug("GPIO Mode: {}".format(
+			GPIO.getmode()
+		))
+		self._logger.info("Light pin: {}, inverted_input: {}, button pin: {}".format(
+			self._settings.get([LIGHT_PIN]),
+			self._settings.get([INVERTED_OUTPUT]),
+			self._settings.get([BUTTON_PIN])
 		))
 		self._logger.info("--------------------------------------------")
 
-		# Setting the default state of pin
-		GPIO.setup(int(self._settings.get(["light_pin"])), GPIO.OUT)
-		if bool(self._settings.get(["inverted_output"])):
-			GPIO.output(int(self._settings.get(["light_pin"])), GPIO.HIGH)
-		else:
-			GPIO.output(int(self._settings.get(["light_pin"])), GPIO.LOW)
+		# Setting the default state of the light pin
+		self.setup_pin()
+		if not bool(self._settings.get([IS_LED_STRIP])):
+			if bool(self._settings.get([INVERTED_OUTPUT])):
+				GPIO.output(int(self._settings.get([LIGHT_PIN])), GPIO.HIGH)
+			else:
+				GPIO.output(int(self._settings.get([LIGHT_PIN])), GPIO.LOW)
+
+		self._plugin_manager.send_plugin_message(self._identifier, dict(isLightOn=self.light_state))
 
-		#Because light is set to ff on startup we don't need to retrieve the current state
-		"""
-		r = self.light_state = GPIO.input(int(self._settings.get(["light_pin"])))
-        if r==1:
-                self.light_state = False
-        else:
-                self.light_state = True
+		# Enabling watch to the default button pin
+		self.enable_watch_button(self._settings.get([BUTTON_PIN]))
 
-        self._logger.info("After Startup. Light state: {}".format(
-                self.light_state
-        ))
-        """
+	def on_api_get(self, request):
+		self._logger.info("Got request. Light state: {}".format(
+			self.light_state
+		))
+
+		if self._settings.get([BUTTON_PIN]) != self._settings.get([PREVIOUS_BUTTON_PIN]):
+			# stop watching on the previous pin
+			GPIO.remove_event_detect(
+				self._settings.get([PREVIOUS_BUTTON_PIN]))
+			# enable watching on the new pin
+			self.enable_watch_button(self._settings.get([BUTTON_PIN]))
+			self._settings.set([PREVIOUS_BUTTON_PIN],
+							   self._settings.get([BUTTON_PIN]))
+
+		self.setup_pin()
+
+		self.change_light_state(None)
+
+		return flask.jsonify(status="ok")
 
+	def on_event(self, event, payload):
 		self._plugin_manager.send_plugin_message(self._identifier, dict(isLightOn=self.light_state))
+		if event == Events.CLIENT_OPENED:
+			return
 
-	def on_api_get(self, request):
-		# Sets the GPIO every time, if user changed it in the settings.
-		GPIO.setup(int(self._settings.get(["light_pin"])), GPIO.OUT)
 
+	def enable_watch_button(self, button):
+		self._logger.info("watching events on pin : {}".format(
+			button
+		))
+		GPIO.setup(button, GPIO.IN, pull_up_down=GPIO.PUD_UP)
```

Comment:
@emouty you need to wrap button in this line and below (or in L96 and 108) with int()

otherwise an exception is thrown

```
octoprint.server.api - ERROR - Error calling SimpleApiPlugin octolight
Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/server/api/__init__.py", line 94, in pluginData
    response = api_plugin.on_api_get(request)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/util/__init__.py", line 1737, in wrapper
    return f(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_octolight/__init__.py", line 108, in on_api_get
    self.enable_watch_button(self._settings.get([BUTTON_PIN]))
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_octolight/__init__.py", line 128, in enable_watch_button
    GPIO.setup(button, GPIO.IN, pull_up_down=GPIO.PUD_UP)
ValueError: Channel must be an integer or list/tuple of integers